lib/models/unfurl.js

```diff
@@ -0,0 +1,226 @@
+};
+
+module.exports = (sequelize, DataTypes) => {
+  const { SlackWorkspace, SlackUser, GitHubUser } = sequelize.models;
+  const Unfurl = sequelize.define('Unfurl', {
+    url: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    channelSlackId: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    githubType: {
+      type: DataTypes.STRING,
+      allowNull: true,
+    },
+    isCondensed: {
+      type: DataTypes.BOOLEAN,
+      allowNull: false,
+    },
+    isPublic: {
+      type: DataTypes.BOOLEAN,
+      allowNull: false,
+    },
+    isDelivered: {
+      type: DataTypes.BOOLEAN,
+      allowNull: false,
+    },
+    slackMessageTimestamp: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+  });
+
+  Unfurl.associate = (models) => {
+    Unfurl.belongsTo(models.SlackWorkspace, {
+      foreignKey: 'slackWorkspaceId',
+      allowNull: false,
+    });
+    Unfurl.belongsTo(models.SlackUser, {
+      foreignKey: 'slackUserId',
+      allowNull: true,
+    });
+  };
+
+
+  Object.assign(Unfurl.prototype, {
+    async removeUnfurlEligibility() {
+      const slackWorkspace = await SlackWorkspace.findById(this.slackWorkspaceId);
+      const key = `${slackWorkspace.slackId}-${this.channelSlackId}-${this.url}`;
+      return cache.set(key, true);
+    },
+
+    async deliver() {
+      const slackUser = await SlackUser.findById(this.slackUserId);
+      const github = await this.constructor.getGitHubClient(this.channelSlackId, slackUser.slackId);
+      const {
+        attachment,
+        githubType,
+      } = await this.constructor.getAttachment(github, this.url, this.isCondensed, this.isPublic);
+      const unfurls = {
+        [this.url]: attachment,
+      };
+      logger.debug(unfurls, 'Unfurling links');
+      const slackWorkspace = await SlackWorkspace.findById(this.slackWorkspaceId);
+      const slackRes = await slackWorkspace.client.chat.unfurl({
+        ts: this.slackMessageTimestamp,
+        channel: this.channelSlackId,
+        unfurls,
+      });
+      logger.trace(slackRes, 'Unfurl complete');
+      await this.removeUnfurlEligibility();
+      this.githubType = githubType;
+      this.isDelivered = true;
+      await this.save();
+    },
+  });
+
+  Object.assign(Unfurl, {
+    async getGitHubClient(channelId, userSlackId) {
+      const github = new GitHubApi({ logger });
+
+      if (!process.env.ALLOWED_CHANNELS || !process.env.ALLOWED_CHANNELS.split(',').includes(channelId)) {
+        // keep the old way of doing things working
+        if (process.env.GITHUB_TOKEN) {
+          github.authenticate({
+            type: 'token',
+            token: process.env.GITHUB_TOKEN,
+          });
+          logger.debug({
+            channelId,
+          }, 'using PAT authed github client for unfurl');
+        }
+        return github;
+      }
+
+      const slackUser = await SlackUser.findOne({
+        where: { slackId: userSlackId },
```

Comment:
Does this also need to scope based on workspace?

Reply:
Yep 👍 Changing this to use our internal id 😬